src/Hook.php

```diff
@@ -0,0 +1,105 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Roave\YouAreUsingItWrong;
+
+use Composer\Composer;
+use Composer\EventDispatcher\EventSubscriberInterface;
+use Composer\IO\IOInterface;
+use Composer\Plugin\PluginInterface;
+use Composer\Script\Event;
+use Composer\Script\ScriptEvents;
+use PackageVersions\Versions;
+use Psalm\Internal\Analyzer\ProjectAnalyzer;
+use Psalm\Internal\Provider\FileProvider;
+use Psalm\Internal\Provider\Providers;
+use Psalm\IssueBuffer;
+use Roave\YouAreUsingItWrong\Composer\Project;
+use Roave\YouAreUsingItWrong\Psalm\Configuration;
+use Roave\YouAreUsingItWrong\Psalm\ProjectFilesToBeTypeChecked;
+use RuntimeException;
+use function define;
+use function defined;
+use function getcwd;
+use function microtime;
+
+/** @internal this is a composer plugin: do not rely on it in your sources */
+final class Hook implements PluginInterface, EventSubscriberInterface
+{
+    private const THIS_PACKAGE_NAME = 'roave/you-are-using-it-wrong';
+
+    /**
+     * {@inheritDoc}
+     */
+    public function activate(Composer $composer, IOInterface $io) : void
+    {
+        // Nothing to do here, as all features are provided through event listeners
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public static function getSubscribedEvents() : array
+    {
+        return [
+            ScriptEvents::POST_INSTALL_CMD => 'runTypeChecks',
+            ScriptEvents::POST_UPDATE_CMD  => 'runTypeChecks',
+        ];
+    }
+
+    /**
+     * @throws RuntimeException
+     */
+    public static function runTypeChecks(Event $composerEvent) : void
+    {
+        $io       = $composerEvent->getIO();
+        $composer = $composerEvent->getComposer();
+        $project  = Project::fromComposerInstallationContext($composer->getPackage(), $composer->getLocker(), getcwd());
+
+        if (! $project->strictTypeChecksAreEnforcedByLocalInstallation()) {
+            $io->write('<info>' . self::THIS_PACKAGE_NAME . ':</info> plugin not installed locally - skipping type checks...');
+
+            return;
+        }
+
+        if ($project->alreadyHasOwnPsalmConfiguration()) {
+            $io->write('<info>' . self::THIS_PACKAGE_NAME . ':</info> psalm.xml detected - skipping static analysis, up to project maintainer to run it...');
```

Comment:
Is there a good reason for writing this output message? I think it could be confusing and alarming, especially for someone who doesn't know about this package but is already running psalm and would wonder why static analysis is being skipped.

![Screenshot from 2019-05-04 11-01-59](https://user-images.githubusercontent.com/159481/57177537-9e191000-6e5c-11e9-93e8-84dd7ee0b73f.png)

Reply:
This line is intentional, and intended to sound like a warning. In your output, the analysis still did run though: that is unexpected.

Maybe there's a better way to phrase this message?

Reply:
The analysis ran because my own `build.sh` directly calls ./vendor/bin/psalm. That seems like a fairly common scenario, and one where I don't need a warning.

Reply:
Oh, that's fine, but as someone mentioned on twitter, you could just `touch psalm.xml` and call it a day. I did want to make this explicit somehow, but thr message I picked seems to be perceived incorrectly, so it needs rephrasing.

Reply:
I'd have a go at coming up with a new phrasing but I'm not clear on the purpose of the message. Aren't you assuming that the project maintainer knows what they're doing if psalm.xml exists? In which case what do you need to warn them about? You can warn them that `you-are-using-it-wrong` will not perform S.A., but since they probably weren't expecting it to anyway I'm not sure that's helpful.

Reply:
There needs to surely be some sort of message indicating that the process bailed out on its own. Lemme know if you can think of a better sentence, otherwise we can defer it to a later issue/pr.

Reply:
hmm, not sure. Maybe something like:

```suggestion
            $io->write('<info>' . self::THIS_PACKAGE_NAME . ':</info>psalm.xml detected - assuming static analysis will run later; not running psalm now');
```

Reply:
Thanks, applied!